docs/docs/loading-data-asynchronously.md

```diff
@@ -0,0 +1,208 @@
+      fetch(`https://api.github.com/users/${username}/gists`)
+        .then(res => res.json())
+        .then(gists => this.setState({ gists }));
+    }
+  }
+
+  render() {
+    const { username } = this.props;
+    const { gists } = this.state;
+    return (
+      <div>
+        <h1>Gists by {username}.</h1>
+        {gists.map(gist => 
+          <p><a href={gist.html_url}>{gist.id}</a></p>
+        )}
+      </div>
+    );
+  }
+}
+```
+
+We can extract the common code in `componentDidMount` and `componentDidUpdate` into a new method, `fetchGists`, and call that in both lifecycle hooks.
+
+```javascript{8,13,17-22}
+class Gists extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {gists: []};
+  }
+
+  componentDidMount() {
+    this.fetchGists();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.username !== prevProps.username) {
+      this.fetchGists();
+    }
+  }
+
+  fetchGists() {
+    const { username } = this.props;
+    fetch(`https://api.github.com/users/${username}/gists`)
+      .then(res => res.json())
+      .then(gists => this.setState({ gists }));
+  }
+
+  render() {
+    const { username } = this.props;
+    const { gists } = this.state;
+    return (
+      <div>
+        <h1>Gists by {username}.</h1>
+        {gists.map(gist => 
+          <p><a href={gist.html_url}>{gist.id}</a></p>
+        )}
+      </div>
+    );
+  }
+}
+```
+
+[Try it out on CodePen.](http://codepen.io/rthor/pen/kkqrQx?editors=0010)
+
+We can simplify the `fetchGists` method by using the [`async / await`](https://tc39.github.io/ecmascript-asyncawait/) feature:
+
+```javascript{1,3-4}
+async fetchGists() {
+  const { username } = this.props;
+  const data = await fetch(`https://api.github.com/users/${username}/gists`);
+  this.setState({gists: await data.json()});
+}
+```
+
+[Try it out on CodePen.](https://codepen.io/rthor/pen/xEoWod?editors=0010)
+
+> **Note:**
+> 
+> `async / await` is still a proposal for the ECMAScript spec and therefore hasn't been implemented in most browsers. To use it today, a [Babel](http://babeljs.io/docs/plugins/transform-async-to-generator/) (or similar) transform is needed. If you're using Create React App, it works by default.
+
+## Pitfalls
+
+An old promise can be pending when a newer promise fulfills. This can cause the old promise to override the result of the new one. If a promise is pending when a component is updated, the result of the first promise should be ignored before a new one is created.
+
+Additionally, a component can unmount while a promise is pending. React warns you if you call `setState` on unmounted components to prevent memory leaks. Some data fetching APIs allow you to cancel requests, and this is preferable when a component unmounts. For APIs such as `fetch` that don't offer a cancellation mechanism, you need to keep track of whether the component is mounted to avoid seeing warnings. Here is how we could implement this:
```

Comment:
I'm not sure what you mean by resolving the unmounting issue as it can be a completely correct behaviour and has nothing to do with the request itself. 

> API calls should be happening as high up in your component tree as possible.

I would argue against this approach. If data is passed down the entire tree, without defining `componentShouldUpdate` at every step, the entire app will be rerendering with every response.